c_src/exml_event.c

```diff
@@ -16,49 +15,79 @@ static ERL_NIF_TERM encode_name(expat_parser *parser_data, const XML_Char *name)
                 {
                     name_len = prefix_start - name_start;
                     prefix_len = strlen(prefix_start+1);
-                    unsigned char *data =  enif_make_new_binary(parser_data->env, name_len+prefix_len, &encoded);
-                    strncpy((char*)data, prefix_start+1, prefix_len);
-                    strncpy((char*)data+prefix_len, name_start, name_len);
-                    data[prefix_len] = ':';
+                    enif_alloc_binary(name_len + prefix_len, &encoded);
+                    strncpy((char*)encoded.data, prefix_start + 1, prefix_len);
+                    strncpy((char*)encoded.data + prefix_len, name_start, name_len);
+                    encoded.data[prefix_len] = ':';
                 } else
                 {
                     name_len = strlen(name_start+1);
-                    unsigned char *data = enif_make_new_binary(parser_data->env, name_len, &encoded);
-                    strncpy((char*)data, name_start+1, name_len);
+                    enif_alloc_binary(name_len, &encoded);
+                    strncpy((char*)encoded.data, name_start + 1, name_len);
                 }
         } else
         {
             size_t name_len = strlen(name);
-            unsigned char* data = enif_make_new_binary(parser_data->env, name_len, &encoded);
-            strncpy((char*)data, name, name_len);
+            enif_alloc_binary(name_len, &encoded);
+            strncpy((char*)encoded.data, name, name_len);
         };
 
     return encoded;
 };
 
+static void unset_handlers(XML_Parser parser)
+{
+    XML_SetStartElementHandler(parser, NULL);
+    XML_SetEndElementHandler(parser, NULL);
+    XML_SetCharacterDataHandler(parser, NULL);
+    XML_SetStartNamespaceDeclHandler(parser, NULL);
+}
+
+static long get_current_byte_index(XML_Parser parser)
+{
+    long index = XML_GetCurrentByteIndex(parser);
+    return index < 0 ? 0 : index;
+}
+
 static void *start_element_handler(expat_parser *parser_data, const XML_Char *name, const XML_Char **atts)
 {
+    ErlNifBinary encoded_name = encode_name(parser_data, name);
     ERL_NIF_TERM element_name;
-    ERL_NIF_TERM attrs_list = enif_make_list(parser_data->env, 0);
+    ERL_NIF_TERM attrs_list;
     int i;
 
-    element_name = encode_name(parser_data, name);
+    if (!parser_data->is_stream_start && parser_data->start_tag != NULL &&
```

Comment:
If I understand it correctly this `if` decides if the parser can be automatically restarted, am I right?

Could this condition be made more readable and easier to follow? By for instance introducing a function with some meaningful name?